Improve parallel iterators splitting behaviour

WaDelma opened the pull request:
Improves parallel iterators splitting behaviour and refactors it.

The improvement:
Before 1000000011111111 was split to 10000000 and to 11111111
After 10001111 is split to 10000000111 and to 11111

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/slide-rs/hibitset/19)
<!-- Reviewable:end -->

WaDelma committed:
Initial work for improving parallel iterator splitting

WaDelma committed:
Refactoring and comments

WaDelma committed:
Better explanation what the splitting algorithm does and improved documentation.

torkleyy commented:
This looks really interesting! I can't say when I have time to review it, in case I forget it feel free to ping me.

torkleyy commented:
@csherratt I would really appreciate if you could also take a look at this PR.
@WaDelma Given that I'm very busy at the moment, it's possible that I can't review this until Friday or Saturday.

WaDelma commented:
Np, I am not in hurry!

torkleyy commented on src/util.rs:
```diff
@@ -55,3 +55,323 @@ impl Row for Index {
+}
+
+#[allow(dead_code)]
+fn average_ones_u64(n: u64) -> Option<u64> {
```
This is hard for me to understand

WaDelma replied:
It basically does parallel [bit count](http://bits.stephan-brumme.com/countBits.html) and then finds place where the bit count goes over half using branchless binary search.

torkleyy replied:
After some benchmarking, it seems that the parallel bit counting is slower than std's `count_ones`. Additionally, a very dumb search is faster than the branchless binary search. So in total we can cut the time in half.

torkleyy replied:
Hold on, for larger numbers it's the exact opposite.

WaDelma replied:
Interesting... Dunno if the speed is worth it for hard to understandness.
Splitting is not even that performance critical..

WaDelma replied:
I can also do branchful binary search.

torkleyy reviewed (commented):
I couldn't spot an issue so far, but I didn't understand all of it yet (especially the `average_ones`). Also, did you benchmark the splitting? Would be interesting to see the difference.

torkleyy commented on src/iter/parallel.rs:
```diff
@@ -109,6 +113,44 @@ impl<'a, T: 'a + Send + Sync> UnindexedProducer for BitProducer<'a, T>
+    let mut count = 0;
+    visit(us, 2, 6, "u".to_owned(), &mut count);
+    visit(them.expect("Splitting top level"), 2, 6, "t".to_owned(), &mut count);
+    assert_eq!(524288 / 64, count);
```
You sure this is true for 32-bit?

WaDelma replied:
Ah that's a mistake!

WaDelma commented:
I would do benchmarking, but my main PC died and I am using crummy laptop.

torkleyy commented on src/iter/parallel.rs:
```diff
@@ -44,56 +43,61 @@ impl<'a, T: 'a + Send + Sync> UnindexedProducer for BitProducer<'a, T>
+                let first_bit = self.0.masks[level].trailing_zeros();
+                average_ones(self.0.masks[level])
+                    .and_then(|average_bit| {
+                        let mask = (1 << (average_bit - 1)) - 1;
```
If you don't know that `average_bit` points to the bit that isn't in the less significant part anymore, this is confusing.

WaDelma replied:
I am bit confused what you mean by this comment...

kvark replied:
what guarantees that `average_bit != 0`?

torkleyy replied:
`average_bits` won't return `0`

WaDelma replied:
This is in else branch of `self.0.masks[level] == 0`, so `average_bit` won't be 0.

WaDelma replied:
Now I realised what you meant by that first comment @torkleyy!

WaDelma committed:
Max maximum splitting test work with 32 bit systems too.

WaDelma committed:
Branching binary search.

torkleyy commented on src/util.rs:
`````diff
@@ -55,3 +55,329 @@ impl Row for Index {
+/// ````
+// TODO: Can 64/32 bit variants be merged to one implementation?
+pub fn average_ones(n: usize) -> Option<usize> {
+    #[cfg(target_pointer_width= "64")]
`````
You can also do conditional compilation on statements like this

```rust
fn average_ones(n: usize) -> Option<usize> {
    #[cfg(..)]
    let res = average_ones_u64(n as u64).map(|n| n as usize);
    // ..
}
```

WaDelma replied:
Ah that's stable? Nice.

torkleyy commented on src/util.rs:
```diff
@@ -55,3 +55,329 @@ impl Row for Index {
+    ];
+
+    // Counting set bits in parallel
+    let a = n - ((n >> 1) & PAR[0]);
```
You can just use the `count_ones` method here

WaDelma replied:
I can't use it as I need the partial results of the counting for the binary search. If this wasn't the case I would have used it in a heart beat.

torkleyy replied:
Ah okay, makes sense.

torkleyy reviewed (commented):
Please put all tests in a `tests` module with a `test` cfg attribute.

WaDelma committed:
Fixed and improved conditional compilation, moved tests to modules.

torkleyy commented:
This PR looks good to me, thank you! Would still be nice if @kvark or @csherratt could take a look.

---

Reviewed 2 of 2 files at r4.
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/slide-rs/hibitset/19#-:-KzyHKyzAtsjP1nKAFys:b-i51riv)*
<!-- Sent from Reviewable.io -->

kvark commented:
What difference does the benchmark show? If not, we should have a benchmark that does.

kvark commented on src/iter/parallel.rs:
```diff
@@ -44,56 +43,61 @@ impl<'a, T: 'a + Send + Sync> UnindexedProducer for BitProducer<'a, T>
-                // The lower half starts iterating from the first bit
-                self.0.prefix[level - 1] = (level_prefix | first_bit) << BITS;
-                Some(other)
+                let first_bit = self.0.masks[level].trailing_zeros();
```
please note that for performance reasons (and this is a perf-critical code, if I understand correctly) we should be using unsafe getters in order to avoid bound checks for `[level]`

kvark replied:
A somewhat cleaner solution would be having each level with a struct (i.e. array of structs instead of structure of arrays `masks`, `prefix`, etc), and then pass references to the whole level around (previous and the current one).

torkleyy replied:
I don't think we need to use unsafe here, it's inside a closure and the index should really get inlined.

WaDelma replied:
As @torkleyy said this is called within closure and after inlining the `level` is a constant value.
Also `masks`, `prefix` are from `BitIter` so editing those would also change sequential iteration code.

torkleyy commented:
Ah right, we need a benchmark!

kvark commented on src/iter/parallel.rs:
```diff
@@ -44,56 +43,61 @@ impl<'a, T: 'a + Send + Sync> UnindexedProducer for BitProducer<'a, T>
+                        Some(other)
+                    }).or_else(|| {
+                        // Because there is only one bit left we descend to it
+                        let idx = level_prefix as usize | first_bit as usize;
```
this logic just becomes one giant magic blob to me, tbh
We need more in-code documentation diagrams and explanations to make this maintenable

torkleyy replied:
What exactly do you think is magic?

WaDelma replied:
What is the hardest thing to understand from it? I have been reading/writing the code, so I am not sure what to improve.

kvark replied:
it's just a solid piece of bit-mangling code
my old brain requires extra focus/energy to go through it :/

torkleyy commented:
I asked for one first, but forgot it then.

kvark commented on src/util.rs:
```diff
@@ -55,3 +55,332 @@ impl Row for Index {
+}
+
+#[allow(dead_code)]
+fn average_ones_u32(n: u32) -> Option<u32> {
```
why is this `dead_code` as opposed to straight be gated by `target_pointer_width`?

WaDelma replied:
Because it's used by tests. We want to run the tests for both variants.

It could be possible gate it with tests too though...

kvark replied:
please, yes, let's do `cfg(test)`

kvark commented on src/util.rs:
```diff
@@ -55,3 +55,332 @@ impl Row for Index {
+    }
+
+    #[test]
+    fn parity_1_average_ones_u64() {
```
pretty nice tests!

kvark commented on src/util.rs:
```diff
@@ -55,3 +55,332 @@ impl Row for Index {
+            }
+        }
+    }
+}
```
nit: newline

WaDelma committed:
Fixed off-by-one error in example. Refactored substraction into the average ones function. Added comments.

WaDelma commented:
Benchmark results on AMD Ryzen 5 1600 (6 cores / 12 SMT x 3.2 GHz)

par_n is parallel iterator with splitting of n-highest levels.

There seems to be quite bit of overhead for parallelism, but after if there is enough bits it's more efficient.

I didn't notice any change between the old splitting strategy and the new one, but the new one should work better if the bitset is unevenly distributed. (I should probably figure out benchmark for that).

  | seq | Par_3 | Par_2 | Par_1
-- | -- | -- | -- | --
100 | 690 | 19850 | 19045 | 9721
1000 | 5611 | 34946 | 27457 | 12587
10000 | 48782 | 56505 | 34214 | 32321
100000 | 331948 | 101915 | 66843 | 159194
1000000 | 1162002 | 234203 | 201314 | 614811

log-log scale
![iter benchmarks](https://i.imgur.com/XOISrQC.png)
linear scale
![iter benchmarks](https://i.imgur.com/sNl3xmH.png)

WaDelma committed:
Iteration benchmarks with and without payload.

Made it possible to configure how many splits are done when doing parallel iteration.

WaDelma commented:
After sleeping night I also realised that that benchmark is flawed: The per iteration computation is minimal, which means that it's measuring mostly the overhead of rayon.

I need to make benchmark with more realistic per iteration computation as it should make splitting on all layers more efficient. I should also probably benchmark it with unbalanced per iteration computation too.

WaDelma commented:
I benchmarked different levels of splitting with different payloads.
It seems that if there is per entity work then there is less difference between how many layers splitting is done.
I reverted it to split to all 3 levels as it seems more stable and efficient when you have lots of entities (and I think in some rayon blog it was mentioned that you should split as much as you can... Not sure about this though).

Benchmarking is annoying. I just realised I should probably also benchmark with different amount of cores.

  | Par_2 | Par_3 | Par_2_payload_100 | Par_3_payload_100 | Par_2_payload_1000 | Par_3_payload_1000 | Par_2_payload_10000 | Par_3_payload_10000
-- | -- | -- | -- | -- | -- | -- | -- | --
100 | 19048 | 20035 | 20771 | 21419 | 25701 | 30633 | 87458 | 126330
1000 | 27744 | 35203 | 34103 | 43724 | 130098 | 100184 | 1063586 | 647348
10000 | 34135 | 57760 | 142638 | 135950 | 1050928 | 684282 | 6240367 | 6146004
100000 | 67404 | 100416 | 800828 | 818030 | 6189107 | 6077761 | 59545648 | 58536681
1000000 | 200724 | 235456 | 7727451 | 5184305 | 41417368 | 40561180 | 403868928 | 393994208

![Parallel payload iteration benchmark](https://i.imgur.com/cJl9lNX.png)

torkleyy commented:
@WaDelma what's the status of this PR?

---

Review status: 0 of 3 files reviewed at latest revision, 5 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/slide-rs/hibitset/19#-:-L0zKUSQQZivdU_RXUGM:b2x1p4f)*
<!-- Sent from Reviewable.io -->

WaDelma commented:
As long as you are happy with my benchmarking and don't find anything else wrong with the PR, I think this should be ready.

torkleyy reviewed (approved):
LGTM, thanks!

torkleyy commented:
@kvark Are you happy with this PR? Or do we still need some more diagrams and comments?

kvark commented on benches/iter.rs:
```diff
@@ -0,0 +1,303 @@
+}
+
+#[bench]
+fn par_payload_1000_iter_10000(b: &mut Bencher) {
```
why aren't we sharing the implementations of those benches, given they only differ in a couple of numbers?

kvark commented on src/iter/parallel.rs:
```diff
@@ -44,56 +43,65 @@ impl<'a, T: 'a + Send + Sync> UnindexedProducer for BitProducer<'a, T>
+    /// 1) First the highest layer that has at least one set bit
+    ///    is searched.
+    ///
+    /// 2) If the layer that was found only has one set bit,
```
only has -> has only

kvark reviewed (commented):
Thanks for the benchmarks!
The advantage of par3 over par2 is not strong, and I could see the appeal to split less given a simpler control flow.
Can we make it configurable (without duplicating much code) please?

WaDelma committed:
Improved documentation.

Purged Wrapping. Refactored the logic that determines to how many layers to split.

WaDelma commented:
splits | largest smallest job
-- | --
1 | usize^3
2 | usize^2
3 | usize

WaDelma commented:
Whoops... Tried to clear comment and accidentally clicked `Close and comment`.

kvark reviewed (approved):
Great stuff! I think you nailed it this time:)
Any more concerns from other folks? Also, do you want to squash some of the intermediate changes?
Otherwise, let's ship it ;)

torkleyy commented:
As @kvark said it would be nice to have the commits squashed or reorganized. Otherwise this is really great, so please r=me once you enhanced the commits.

bors delegate=WaDelma

WaDelma committed:
Moved average ones behind parallel feature.

WaDelma commented:
Should I remove some of the benchmarks? They seem to slow CI down.

Also I am not that happy with `average_ones` as it's results are bit unintuitive: 01001 is split 01, 001 and not 0100, 1. It doesn't affect it's use, but it makes for weird API and explanation. Dunno if it's worth changing.

kvark commented:
Follow up is fine. As for benches - I don't think we should rely on CI anyway, so we can turn the par iter off there.

> On Jan 2, 2018, at 04:30, WaDelma <notifications@github.com> wrote:
> 
> Should I remove some of the benchmarks? They seem to slow CI down.
> 
> Also I am not that happy with average_ones as it's results are bit unintuitive: 01001 is split 01, 001 and not 0100, 1. It doesn't affect it's use, but it makes for weird API and explanation. Dunno if it's worth changing.
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or mute the thread.
>

WaDelma committed:
Feature gated parallel benchmarks and made travis only bench non-parallel ones.

WaDelma commented:
r=me

WaDelma commented:
bors r=me

torkleyy commented:
@WaDelma You can approve a PR for me by writing "bors r=torkleyy".

WaDelma commented:
ah, sorry. Don't really know how bors works :P

WaDelma commented:
bors r-

WaDelma commented:
bors r=torkleyy
